### ISSUE TITLE
ICE when casting float to integer (out-of-bounds)

### ISSUE THREAD

**Issue by samnardoni** [rust-lang/rust#45134]
The following program (with perhaps undefined behaviour) causes an ICE:
```rust
fn ice() {
    println!("{:?}", Some(257.0f32 as u8));
}
```

Interestingly, if not wrapping the result in a `Some`, then there's no ICE.
```rust
fn no_ice() {
    println!("{:?}", 257.0f32 as u8);
}
```

Here's the result from running it on the playground:
```rust
   Compiling playground v0.0.1 (file:///playground)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0 (f3d6973f4 2017-08-27) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'assertion failed: !is_undef(val)', /checkout/src/librustc_trans/mir/constant.rs:1083:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by scottmcm** [comment]
Looks like https://github.com/rust-lang/rust/issues/10184 (which is approaching its fourth birthday)

**Comment by rkruppe** [comment]
Same underlying issue, but a different code path than the one that's been exclusively (as far as I can tell) discussed over there. So this report was very valuable for my in-progress patch addressing #10184, thank you @samnardoni!

**Comment by samnardoni** [comment]
@rkruppe No problem at all. I'm glad it's useful.
